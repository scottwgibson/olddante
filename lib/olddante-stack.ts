import * as cdk from '@aws-cdk/core';
import { StackProps } from '@aws-cdk/core';
import { Function, Runtime, Code } from "@aws-cdk/aws-lambda"
import { LambdaIntegration, MethodLoggingLevel, RestApi } from "@aws-cdk/aws-apigateway"
import { Bucket, BlockPublicAccess} from "@aws-cdk/aws-s3"
import { StringParameter } from '@aws-cdk/aws-ssm';

export class OldDanteStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const bucket = new Bucket(this, this.stackName + "Bucket", {
      versioned: false,
      publicReadAccess: false,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    const lambda = new Function(this, this.stackName + "Lambda", {
      functionName: this.stackName + "Lambda",
      handler: "handler.lambda_handler",
      runtime: Runtime.PYTHON_3_7,
      code: Code.fromAsset('src/lambda.zip'),
      memorySize: 512,
      timeout: cdk.Duration.seconds(60),
      environment: {
        bucketName: bucket.bucketName
      }
    })
    bucket.grantRead(lambda)

    StringParameter.fromStringParameterName(this, 'SlackBotToken', '/olddante/slack/bot_token')
      .grantRead(lambda);

    StringParameter.fromStringParameterName(this, 'SlackSigningSecret', '/olddante/slack/signing_secret')
      .grantRead(lambda)


    const api = new RestApi(this, this.stackName + "RestApi", {
        deployOptions: {
            stageName: "prod",
            metricsEnabled: true,
            loggingLevel: MethodLoggingLevel.INFO,
            dataTraceEnabled: true,
        },
      })
    const integration = new LambdaIntegration(lambda, {})
    api.root.addMethod("GET", integration)
    const slack = api.root.addResource('slack');
    const events = slack.addResource('events')
    events.addMethod("POST", integration)
  }
}
