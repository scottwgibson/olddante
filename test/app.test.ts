import { expect as expectCDK, matchTemplate, MatchStyle } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as App from '../lib/olddante-stack';
import { SynthUtils } from '@aws-cdk/assert';

test('Empty Stack', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new App.OldDanteStack(app, 'MyTestStack');
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
