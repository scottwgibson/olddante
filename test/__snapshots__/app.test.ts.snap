// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty Stack 1`] = `
Object {
  "Outputs": Object {
    "MyTestStackRestApiEndpointFF4CBA1F": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "MyTestStackRestApi739A2852",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "MyTestStackRestApiDeploymentStageprodABCD6224",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0ArtifactHashCADD2773": Object {
      "Description": "Artifact hash for asset \\"348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0\\"",
      "Type": "String",
    },
    "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0S3Bucket78AF96F1": Object {
      "Description": "S3 bucket for asset \\"348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0\\"",
      "Type": "String",
    },
    "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0S3VersionKey6DB321C2": Object {
      "Description": "S3 key for asset version \\"348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0\\"",
      "Type": "String",
    },
    "SlackBotTokenParameter": Object {
      "Default": "/olddante/slack/bot_token",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SlackSigningSecretParameter": Object {
      "Default": "/olddante/slack/signing_secret",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyTestStackBucket8E3B7739": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestStackLambdaE1C4A331": Object {
      "DependsOn": Array [
        "MyTestStackLambdaServiceRoleDefaultPolicy739F0D62",
        "MyTestStackLambdaServiceRole0CC86EF6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0S3Bucket78AF96F1",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0S3VersionKey6DB321C2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters348b645ed665777029bbd238d1ec1f95b0f219594cd596a566e6ae3dd95743c0S3VersionKey6DB321C2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "bucketName": Object {
              "Ref": "MyTestStackBucket8E3B7739",
            },
          },
        },
        "FunctionName": "MyTestStackLambda",
        "Handler": "handler.lambda_handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestStackLambdaServiceRole0CC86EF6",
            "Arn",
          ],
        },
        "Runtime": "python3.7",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestStackLambdaServiceRole0CC86EF6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestStackLambdaServiceRoleDefaultPolicy739F0D62": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackBucket8E3B7739",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestStackBucket8E3B7739",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/olddante/slack/bot_token",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/olddante/slack/signing_secret",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestStackLambdaServiceRoleDefaultPolicy739F0D62",
        "Roles": Array [
          Object {
            "Ref": "MyTestStackLambdaServiceRole0CC86EF6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestStackRestApi739A2852": Object {
      "Properties": Object {
        "Name": "MyTestStackRestApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MyTestStackRestApiAccount86D2457D": Object {
      "DependsOn": Array [
        "MyTestStackRestApi739A2852",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "MyTestStackRestApiCloudWatchRoleC933809B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "MyTestStackRestApiCloudWatchRoleC933809B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestStackRestApiDeploymentC23C6DCE72352015be33c298c8c03590ebe77d6a": Object {
      "DependsOn": Array [
        "MyTestStackRestApiGET833477D3",
        "MyTestStackRestApislackeventsPOSTD005E2F3",
        "MyTestStackRestApislackevents0A81D4F3",
        "MyTestStackRestApislack784D51A5",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MyTestStackRestApiDeploymentStageprodABCD6224": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "MyTestStackRestApiDeploymentC23C6DCE72352015be33c298c8c03590ebe77d6a",
        },
        "MethodSettings": Array [
          Object {
            "DataTraceEnabled": true,
            "HttpMethod": "*",
            "LoggingLevel": "INFO",
            "MetricsEnabled": true,
            "ResourcePath": "/*",
          },
        ],
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "MyTestStackRestApiGET833477D3": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackLambdaE1C4A331",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "MyTestStackRestApi739A2852",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MyTestStackRestApiGETApiPermissionMyTestStackMyTestStackRestApi1CD3A8FAGET50DFD3EA": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestStackLambdaE1C4A331",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "MyTestStackRestApi739A2852",
              },
              "/",
              Object {
                "Ref": "MyTestStackRestApiDeploymentStageprodABCD6224",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestStackRestApiGETApiPermissionTestMyTestStackMyTestStackRestApi1CD3A8FAGETEB6EDBC9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestStackLambdaE1C4A331",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "MyTestStackRestApi739A2852",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestStackRestApislack784D51A5": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "MyTestStackRestApi739A2852",
            "RootResourceId",
          ],
        },
        "PathPart": "slack",
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "MyTestStackRestApislackevents0A81D4F3": Object {
      "Properties": Object {
        "ParentId": Object {
          "Ref": "MyTestStackRestApislack784D51A5",
        },
        "PathPart": "events",
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "MyTestStackRestApislackeventsPOSTApiPermissionMyTestStackMyTestStackRestApi1CD3A8FAPOSTslackevents49F4862D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestStackLambdaE1C4A331",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "MyTestStackRestApi739A2852",
              },
              "/",
              Object {
                "Ref": "MyTestStackRestApiDeploymentStageprodABCD6224",
              },
              "/POST/slack/events",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestStackRestApislackeventsPOSTApiPermissionTestMyTestStackMyTestStackRestApi1CD3A8FAPOSTslackeventsD4BE02D2": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestStackLambdaE1C4A331",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "MyTestStackRestApi739A2852",
              },
              "/test-invoke-stage/POST/slack/events",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestStackRestApislackeventsPOSTD005E2F3": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackLambdaE1C4A331",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "MyTestStackRestApislackevents0A81D4F3",
        },
        "RestApiId": Object {
          "Ref": "MyTestStackRestApi739A2852",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
}
`;
